### PR TITLE
Issue #30: Use config files when initiating session.

### LINKED ISSUE #30: Override session configuration by setting session key in config directory
### Override session configuration by setting session key in config directory

|    Q        |   A
|------------ | ------
| New Feature | yes
| RFC         | no
| BC Break    | no

#### Summary

Modify the constructor of the PhpSessionPersistence.php class to override the session configuration by setting variables in an array under the session key. Why. This would be consistent with being able to set other session parameters....

For example:

```
// file: config/autoload/session.global.php
return [
    'session' => [
		'session.name' => 'own_session_name',
		'session.cookie_secure' => true, 
		'session.cookie_httponly' = true,
		...
        'persistence' => [
            'ext' => [
                'non_locking' => true, // true|false, true => marks read_and_close as true
            ],
        ],
    ],
];
```

### PR DESCRIPTION
**Pull request by alexmerlin**
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | yes
| RFC           | no
| QA            | no

### Description
As discussed in #30 session details should be configurable and use `ini_get` only as a fallback.
This PR modifies `src/PhpSessionPersistence.php` to first look for config values and if not found, use `ini_get` fallbacks.

### CODE REVIEW

File: src/PhpSessionPersistence.php [hunk 1]
```diff
@@ -59,34 +63,55 @@ class PhpSessionPersistence implements InitializePersistenceIdInterface, Session
      * Those headers will be added programmatically to the response along with
      * the session set-cookie header when the session data is persisted.
      *
-     * @param bool $nonLocking use the non locking mode during initialization?
-     * @param bool $deleteCookieOnEmptySession delete cookie from browser when session becomes empty?
+     * @param array $session
      */
-    public function __construct(bool $nonLocking = false, bool $deleteCookieOnEmptySession = false)
+    public function __construct(array $session = [])
```
Comment (Ocramius) [inline src/PhpSessionPersistence.php]: This is a major BC break (API signature change)

**Review by Ocramius — CHANGES_REQUESTED** [review]

File: src/PhpSessionPersistence.php (hunk 1 above)
Comment (gsteel) [reply src/PhpSessionPersistence.php]: Possibly introduce a static constructor, i.e. `fromConfigArray` and use that in the factory, leaving the current constructor unchanged.

Comment (alexmerlin) [reply src/PhpSessionPersistence.php]: Implemented static constructor - please check.

File: src/PhpSessionPersistence.php [hunk 2]
```diff
@@ -89,6 +93,63 @@ public function __construct(bool $nonLocking = false, bool $deleteCookieOnEmptyS
         $this->cookieSameSite = ini_get('session.cookie_samesite');
     }
 
+    /**
+     * @param array $session
+     * @return self
+     */
```
Comment (gsteel) [inline src/PhpSessionPersistence.php]:
```suggestion
     * @internal
     * @return self
     */
```

Marking this as internal means we don't have to maintain it for the next major. i.e. we can then move all this to the constructor as you originally proposed.

File: src/PhpSessionPersistence.php [hunk 3]
```diff
@@ -89,6 +93,63 @@ public function __construct(bool $nonLocking = false, bool $deleteCookieOnEmptyS
...
+    public static function fromConfigArray(array $session = [])
```
Comment (gsteel) [inline src/PhpSessionPersistence.php]: Naming nitpick - ideally, this should be something like `$config` or `$sessionConfig`

File: src/PhpSessionPersistence.php [hunk 4]
```diff
@@ -89,6 +93,63 @@ public function __construct(bool $nonLocking = false, bool $deleteCookieOnEmptyS
         $this->cookieSameSite = ini_get('session.cookie_samesite');
     }
 
+    /**
+     * @param array $session
```
Comment (gsteel) [inline src/PhpSessionPersistence.php]: This parameter could be improved with something like:

```suggestion
     * @param array{
     *   cache_limiter?: string,
     *   other-keys?: whatever,
     *   persistence?: array{
     *     ext?: array{
     *       delete_cookie_on_empty_session?: bool,
     *       other-keys?: whatever,
     *     },
     *   },
     * } $session
```

If declared as a `/** @psalm-type SessionConfig = array{...} */` at class level, the array shape can be imported easily anywhere it's needed - example in [laminas-input-filter](https://github.com/laminas/laminas-inputfilter/blob/a79c31bcb58d1041393dd3a7a8187829e4ca7b02/src/InputFilterInterface.php#L14)

**Review by gsteel — CHANGES_REQUESTED** [review]
Thanks @alexmerlin - I think the static constructor needs to be marked internal _(You'll need to baseline `InternalMethod` psalm issues for tests)_.

The other items are nice-to-haves and otherwise it's a 👍 from me.

File: src/PhpSessionPersistence.php (hunk 3 above)
Comment (gsteel) [reply src/PhpSessionPersistence.php]: Sorry, this method should also be marked as `final` and have the return type `self`

Comment (alexmerlin) [reply src/PhpSessionPersistence.php]: Method is now final and argument is called `$sessionConfig`.

File: src/PhpSessionPersistence.php (hunk 2 above)
Comment (alexmerlin) [reply src/PhpSessionPersistence.php]: Method is now final and internal.

File: src/PhpSessionPersistence.php (hunk 4 above)
Comment (alexmerlin) [reply src/PhpSessionPersistence.php]: I created psalm type `SessionConfig`, but I could not fix the below error so I had to psalm suppress `MixedArgumentTypeCoercion` there.

> ERROR: MixedArgumentTypeCoercion - src/PhpSessionPersistenceFactory.php:42:13 - Argument 1 of Mezzio\Session\Ext\PhpSessionPersistence::fromConfigArray expects array{cache_expire?: int, cache_limiter?: string, cookie_domain?: string, cookie_httponly?: bool, cookie_lifetime?: int, cookie_path?: string, cookie_samesite?: string, cookie_secure?: bool, name?: string, persistence?: array{ext?: array{delete_cookie_on_empty_session?: bool, non_locking?: bool}}}, but parent type array<array-key, mixed> provided (see https://psalm.dev/194)
>
>             isset($config['session']) && is_array($config['session']) ? $config['session'] : []
> 
>   The type of isset($config['session']) && is_array($config['session']) ? $config['session'] : [] is sourced from here - vendor/psr/container/src/ContainerInterface.php:20:16
>      * @return mixed Entry.

**Review by gsteel — APPROVED** [review]
LGTM 👍 - Thanks @alexmerlin

File: src/PhpSessionPersistence.php (hunk 4 above)
Comment (gsteel) [reply src/PhpSessionPersistence.php]: It's acceptable to suppress the issue in the factory. Alternatives are validate the array shape in code somewhere, force the type with `/** @psalm-var SessionConfig */`, or expand the baseline - generally speaking, expanding the baseline is preferred in cases like this 👍

**Pull request merged**